### PR TITLE
Update endpoint slices during service import reconciliation

### PR DESCRIPTION
**Pull request by astaticvoid**
*Issue #, if available:*
#21 

*Description of changes:*
Reconcile service import endpoint slices based on changes in Cloud Map

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/controllers/cloudmap_controller.go [hunk 1]
```diff
@@ -208,37 +208,114 @@ func (r *CloudMapReconciler) createAndGetDerivedService(ctx context.Context, svc
 	return r.getDerivedService(ctx, svc.Namespace, svcImport.Annotations[DerivedServiceAnnotation])
 }
 
-func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport *v1alpha1.ServiceImport, cloudMapService *model.Service, svc *v1.Service) error {
+func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport *v1alpha1.ServiceImport, desiredEndpoints []*model.Endpoint, svc *v1.Service) error {
 	existingSlicesList := discovery.EndpointSliceList{}
-	var existingSlices []*discovery.EndpointSlice
 	if err := r.Client.List(ctx, &existingSlicesList,
 		client.InNamespace(svc.Namespace), client.MatchingLabels{discovery.LabelServiceName: svc.Name}); err != nil {
 		return err
```
Comment (runakash) [inline pkg/controllers/cloudmap_controller.go:215]: Error logging.

File: pkg/controllers/cloudmap_controller.go [hunk 2]
```diff
@@ -208,37 +208,114 @@ func (r *CloudMapReconciler) createAndGetDerivedService(ctx context.Context, svc
...
-		for _, slice := range existingSlices {
-			if err := r.Client.Create(ctx, slice); err != nil {
+
+	desiredPorts := extractEndpointPorts(desiredEndpoints)
+	matchedEndpoints := make(map[string]*discovery.Endpoint)
+	endpointsToCreate := make([]discovery.Endpoint, 0)
+
+	// check if all desired endpoints are in an endpoint slice already
+	for _, desiredEndpoint := range desiredEndpoints {
+		existingEndpoint := findEndpointInSliceByIp(existingSlicesList.Items, desiredEndpoint.IP)
```
Comment (runakash) [inline pkg/controllers/cloudmap_controller.go:224]: This is O(n^2) operation. We can improve here by normalizing `existingSlicesList.Items` as `map[Address]Endpoint` before the for loop, and doing a key lookup.

File: pkg/controllers/cloudmap_controller.go [hunk 3]
```diff
@@ -208,37 +208,114 @@ func (r *CloudMapReconciler) createAndGetDerivedService(ctx context.Context, svc
...
+	for _, existingSlice := range existingSlicesList.Items {
+		updatedEndpointList := make([]discovery.Endpoint, 0)
+		for _, existingEndpoint := range existingSlice.Endpoints {
+			keep, found := matchedEndpoints[existingEndpoint.Addresses[0]]
+			if found {
+				updatedEndpointList = append(updatedEndpointList, *keep)
+			}
+		}
+
+		dirty := len(existingSlice.Endpoints) != len(updatedEndpointList)
```
Comment (runakash) [inline pkg/controllers/cloudmap_controller.go:242]: how about renaming this as `isEndpointSliceUpdated`

File: pkg/controllers/cloudmap_controller.go [hunk 4]
```diff
@@ -208,37 +208,114 @@ func (r *CloudMapReconciler) createAndGetDerivedService(ctx context.Context, svc
...
+
+		// fill endpoint slice with endpoints to create if necessary and there is sufficient room
+		for _, endpointToCreate := range endpointsToCreate {
+			if len(updatedEndpointList) >= 100 {
```
Comment (runakash) [inline pkg/controllers/cloudmap_controller.go:246]: We can have upto [1000](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#endpointslice-resource) endpoints per endpoint slice by configuring `--max-endpoints-per-slice` flag. Are we planning to support this in the future.
Maybe, we can also add comment about why this magic number is used. And extract it out as a constant.

File: pkg/controllers/cloudmap_controller.go [hunk 5]
```diff
@@ -208,37 +208,114 @@ func (r *CloudMapReconciler) createAndGetDerivedService(ctx context.Context, svc
...
+		}
+
+		sliceToUpdate := existingSlice
+		sliceToUpdate.Endpoints = updatedEndpointList
+
+		// delete empty endpoint slice
+		if len(updatedEndpointList) == 0 {
+			r.Logger.Info("deleting EndpointSlice", "namespace", sliceToUpdate.Namespace, "name", sliceToUpdate.Name)
+			if err := r.Client.Delete(ctx, &sliceToUpdate); err != nil {
+				return err
```
Comment (runakash) [inline pkg/controllers/cloudmap_controller.go:261]: Error logging

File: pkg/controllers/cloudmap_controller.go [hunk 6]
```diff
@@ -208,37 +208,114 @@ func (r *CloudMapReconciler) createAndGetDerivedService(ctx context.Context, svc
...
+		// needsUpdate = true if ports don't match
+		if !EndpointPortsAreEqualIgnoreOrder(desiredPorts, sliceToUpdate.Ports) {
+			sliceToUpdate.Ports = desiredPorts
+			dirty = true
+		}
+
+		if dirty {
+			r.Logger.Info("updating EndpointSlice", "namespace", sliceToUpdate.Namespace, "name", sliceToUpdate.Name)
+			if err := r.Client.Update(ctx, &sliceToUpdate); err != nil {
 				return err
```
Comment (runakash) [inline pkg/controllers/cloudmap_controller.go:275]: Error logging

File: pkg/controllers/cloudmap_controller.go [hunk 7]
```diff
@@ -252,37 +329,31 @@ func createDerivedServiceStruct(svc *model.Service, svcImport *v1alpha1.ServiceI
...
+		},
 	}
+}
 
-	// TODO split slices in case there are more than 1000 endpoints
-	// see https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0752-endpointslices/README.md
-
-	slices = append(slices, &discovery.EndpointSlice{
+func createEndpointSliceStruct(svcImport *v1alpha1.ServiceImport, svc *v1.Service, endpoints []discovery.Endpoint, ports []discovery.EndpointPort) *discovery.EndpointSlice {
+	return &discovery.EndpointSlice{
```
Comment (runakash) [inline pkg/controllers/cloudmap_controller.go:356]: One TODO would be add the label `endpoints.kubernetes.io/over-capacity` when we reach 1000 endpoints.

**Review by runakash — CHANGES_REQUESTED** [review]

File: pkg/controllers/cloudmap_controller.go (hunk 2 above)
Comment (astaticvoid) [reply pkg/controllers/cloudmap_controller.go:224]: It's a good idea.

File: pkg/controllers/cloudmap_controller.go (hunk 1 above)
Comment (astaticvoid) [reply pkg/controllers/cloudmap_controller.go:215]: errors are returned and logged in reconcile namespace, so we'd be logging them twice.

Comment (astaticvoid) [reply pkg/controllers/cloudmap_controller.go:215]: I can wrap them in some context.

File: pkg/controllers/cloudmap_controller.go (hunk 4 above)
Comment (astaticvoid) [reply pkg/controllers/cloudmap_controller.go:246]: I don't know if slices that large would be preferable over the previous endpoints api then, or if we'd gain anything by doing that. I think it's something we can explore during load test, make it optionally configurable. I just took the default recommended slice size from the docs, so I don't have an explanation to add, but I agree it should be a constant at least.

File: pkg/controllers/cloudmap_controller.go (hunk 7 above)
Comment (astaticvoid) [reply pkg/controllers/cloudmap_controller.go:356]: I believe since we're using slices, and keeping them small we should not have this problem.

**Review by runakash — APPROVED** [review]
LGTM

**Pull request merged**